components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,648 @@
+            && self.catch_up_logs_ready(c)
+        {
+            let c = self.merge_context_mut().catch_up_logs.take().unwrap();
+            if c.tx.send(()).is_err() {
+                error!(
+                    self.logger,
+                    "failed to respond to merge target, are we shutting down?"
+                );
+                // TODO: rollback.
+            }
+            self.take_merge_context();
+            self.mark_for_destroy(None);
+            return;
+        }
+        self.on_check_merge(store_ctx);
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    // Match v1::on_ready_commit_merge.
+    pub fn on_apply_res_commit_merge<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        res: CommitMergeResult,
+    ) {
+        let tablet: EK = match res.tablet.downcast() {
+            Ok(t) => *t,
+            Err(t) => unreachable!("tablet type should be the same: {:?}", t),
+        };
+        {
+            let mut meta = store_ctx.store_meta.lock().unwrap();
+
+            // In v2 we don't remove source region immediately.
+            let source_region = &meta.regions.get(&res.source.get_id()).unwrap().0;
+            assert!(
+                source_region.get_end_key() == res.region.get_end_key()
+                    || source_region.get_start_key() == res.region.get_start_key()
+            );
+            meta.set_region(&res.region, true, &self.logger);
+            let (reader, read_tablet) = meta.readers.get_mut(&res.region.get_id()).unwrap();
+            self.set_region(
+                &store_ctx.coprocessor_host,
+                reader,
+                res.region.clone(),
+                RegionChangeReason::CommitMerge,
+                res.index,
+            );
+
+            // Tablet should be updated in lock to match the epoch.
+            *read_tablet = SharedReadTablet::new(tablet.clone());
+
+            // After the region commit merged, the region's key range is extended and the
+            // region's `safe_ts` should reset to `min(source_safe_ts, target_safe_ts)`
+            let source_read_progress = meta
+                .region_read_progress
+                .remove(&res.source.get_id())
+                .unwrap();
+            self.read_progress_mut().merge_safe_ts(
+                source_read_progress.safe_ts(),
+                res.index,
+                &store_ctx.coprocessor_host,
+            );
+        }
+        if let Some(tablet) = self.set_tablet(tablet) {
+            self.record_tombstone_tablet(store_ctx, tablet, res.index);
+        }
+        self.record_tombstone_tablet_path(store_ctx, res.source_path, res.index);
+
+        // If a follower merges into a leader, a more recent read may happen
+        // on the leader of the follower. So max ts should be updated after
+        // a region merge.
+        self.txn_context()
+            .on_became_leader(store_ctx, self.term(), self.region(), &self.logger);
```

Comment:
Should add a new method for merge specifically. V1 only needs to update the timestamp.

Reply:
V1 calls the method within lock.